.devcontainer/Makefile

```diff
@@ -0,0 +1,98 @@
+# Copyright 2023 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+srcDir := $(shell pwd)
+buildDir := $(shell pwd)/build
+
+# %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% #
+#     Developer Options - Comment / Uncomment to change Makefile behavior      # 
+# %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% #
+
+# Comment to enable LLVM-CPU
+IREE_LLVM_CPU_DISABLE := 1
+export IREE_LLVM_CPU_DISABLE
+EXCLUDE_LBLS := $(if $(EXCLUDE_LBLS),$(EXCLUDE_LBLS)|,)
+EXCLUDE_LBLS := ${EXCLUDE_LBLS}^driver=llvm-cpu$$
+
+# Comment to enable Vulkan GPU
+IREE_VULKAN_DISABLE := 1
+export IREE_VULKAN_DISABLE
+EXCLUDE_LBLS := $(if $(EXCLUDE_LBLS),$(EXCLUDE_LBLS)|,)
+EXCLUDE_LBLS := ${EXCLUDE_LBLS}^driver=vulkan$$
+
+# Comment to enable FP16 on Vulkan GPU
+IREE_VULKAN_F16_DISABLE := 1
+export IREE_VULKAN_F16_DISABLE
+EXCLUDE_LBLS := $(if $(EXCLUDE_LBLS),$(EXCLUDE_LBLS)|,)
+EXCLUDE_LBLS := ${EXCLUDE_LBLS}^driver=vulkan_uses_vk_khr_shader_float16_int8$$
+
+# Comment to enable NVIDIA GPU
+IREE_CUDA_DISABLE := 1
+export IREE_CUDA_DISABLE
+EXCLUDE_LBLS := $(if $(EXCLUDE_LBLS),$(EXCLUDE_LBLS)|,)
+EXCLUDE_LBLS := ${EXCLUDE_LBLS}^driver=cuda$$
+
+# %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% #
+#                           END of Developer Options                           #
+# %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% #
+
+CTEST_BASE_ARGS := --test-dir ${buildDir} --timeout 900 --output-on-failure --no-tests=error
+CTEST_ARGS := ${CTEST_BASE_ARGS} --label-exclude '(${EXCLUDE_LBLS})'
+
+# The libzstd on Ubuntu 18.04 doesn't work with the IREE integrated LLVM.
+# TODO(#11782): Remove once we bump the minimum supported OS version.
+TEST_1804 = "$(shell lsb_release -d | awk -F"\t" '{print $2}' | grep '18.04')"
+ifneq ("",$(TEST_1804))
+  LLVM_ENABLE_ZSTD_FLAG = -DLLVM_ENABLE_ZSTD=OFF
+endif
+
+CTEST_PARALLEL_LEVEL := $(shell nproc)
+export CTEST_PARALLEL_LEVEL
+
+# %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% #
+#                               Makefile Commands                              # 
+# %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% #
+
+.PHONY: test clean build
+
+build:
+	set -x
+	ninja -C ${buildDir} all
+	set +x
+
+clean:
+	rm -rf ${buildDir}
+
+configure:
+	ccache --max-size=20G
+	mkdir -p ${buildDir}
+	set -x
+	cmake -B ${buildDir} -S . -G Ninja \
+		-DCMAKE_BUILD_TYPE=RelWithDebInfo \
+		-DCMAKE_C_COMPILER=${CC} \
+		-DCMAKE_CXX_COMPILER=${CXX} \
+		-DCMAKE_C_COMPILER_LAUNCHER=ccache \
```

Comment:
Is this Makefile necessary for integration into any particular tool? It looks like a limited subset of our existing CMake build, and I'd prefer to not need to support another build system.

If the CMake build is difficult to configure and use, that would be useful feedback and we can try to streamline it through other means.

Reply:
The `Makefile` per say is not essential. Now the documentation is hard to read. It's hard to find how to do X, Y or Z.
Hard to know which option to use, or even exists. (I still didn't manage to build with NVIDIA GPU support for instance)

The documentation is not doing a very good job here. I made this Makefile to give you the ability to quickly test the feature, and maybe acting as a long term "demo getting started" script.

Reply:
I'll respond to the individual points here, but my goal is definitely to have enough of this be discoverable / self-serve from the documentation.

---

> Hard to know which option to use, or even exists.

Source of truth for this is the root [`CMakeLists.txt`](https://github.com/openxla/iree/blob/main/CMakeLists.txt), though a few flags are listed in [`build_tools/cmake/iree_llvm.cmake`](https://github.com/openxla/iree/blob/main/build_tools/cmake/iree_llvm.cmake). We made an attempt at listing the options in a doc at [`docs/developers/get_started/cmake_options_and_variables.md`](https://github.com/openxla/iree/blob/main/docs/developers/get_started/cmake_options_and_variables.md) (modeled after https://llvm.org/docs/CMake.html), but that page is stale.

> (I still didn't manage to build with NVIDIA GPU support for instance).

The current docs for this are (in order)
* https://openxla.github.io/iree/building-from-source/getting-started/
* CUDA: https://openxla.github.io/iree/deployment-configurations/gpu-cuda-rocm/
* Vulkan: https://openxla.github.io/iree/deployment-configurations/gpu-vulkan/

For CUDA, compiler support is enabled via `-DIREE_HAL_DRIVER_CUDA=ON` (on by default). This will first check for a local CUDA toolkit via `find_package(CUDAToolkit)`, then will fall back to downloading a copy automatically using [`third_party/nvidia_sdk_download/fetch_cuda_toolkit.py`](https://github.com/openxla/iree/blob/main/third_party/nvidia_sdk_download/fetch_cuda_toolkit.py). Runtime support is enabled via `-DIREE_HAL_DRIVER_CUDA=ON` (on by default when the CUDA SDK was found or successfully downloaded).

> It took me a long time to understand how to test the right subset of tests given how you built IREE (and I'm sure I still don't understand the full story, but at least now the tests pass).

We have a script at [`build_tools/cmake/ctest_all.sh`](https://github.com/openxla/iree/blob/main/build_tools/cmake/ctest_all.sh) that uses environment variables to help with filtering. By default, the project currently tries running tests that depend on Vulkan/CUDA/etc. even if your machine does not have the right hardware / drivers: https://github.com/openxla/iree/issues/12042.

> I made this Makefile to give you the ability to quickly test the feature, and maybe acting as a long term "demo getting started" script.

I appreciate the effort (and the feedback!), though I'm not really comfortable directing people to a Makefile for overall project development (what this dev container is set up to be) without some broader discussion and a commitment to supporting it (exercising it by CI). Some other "getting started" pages already exist / are in the works:
* https://openxla.github.io/iree/building-from-source/getting-started/
* https://github.com/openxla/iree/tree/main/samples/dynamic_shapes (note that this sample has a Makefile, but there are caveats in using it)
* https://github.com/benvanik/iree-template-runtime-cmake

Reply:
Another option for viewing the available CMake options is to watch the output of CMake's configure step or to use [`cmake-gui`](https://cmake.org/cmake/help/latest/manual/cmake-gui.1.html) for interactive configuration and search for `IREE_`: 
![image](https://user-images.githubusercontent.com/4010439/230179584-5a2fe7fa-e4ea-4c4e-ae19-af594ddaac2b.png)

Reply:
Yeah I'm also hesitant about adding makefile, especially as the place we direct people. We already support two build systems and I really don't want a third, especially since CMake is supposed to be a layer on top of Make (but cross-platform). I would definitely be interested in configuration or scripting to aid with common setups though. We've discussed CMake presets before. Easier selection of tests would also help.

Reply:
To be fair, the Makefile in this PR calls CMake, so it inherits many of the properties of CMake (portable across Linux/macOS/Windows, configurable, uses multiple cores by default via Ninja, etc.).

Reply:
@GMNGeoffrey @ScottTodd I think the point remains. It doesn't really matter how it's being done.
It would be a good idea to provide some "default scripts or commands" that the user can execute without having to think about it. 

There are a ton of scripts in  https://github.com/openxla/iree/tree/main/build_tools/scripts. Unfortunately nothing as dummy proof / simple as:
- `make configure`
- `make build`
- `make test`
- `make release`
- `make release-android`
- etc.

It's a good opportunity to rethink a little bit the strategy with regards to documentation / scripts. Lowering the bar to entry for developers will not only improve the number of community PRs, but also help new hires & interns of companies contributing to the project to ramp up and become an active contributor.

I understand that "to each their workflow", however we can provide a "common ground" good enough for most people, and let people customize in their own environment. And this can be pretty easily achieve if we come to the agreement of updating the `docker-compose.yml` & `devcontainer.json` as little as possible.

I'm not attached specifically to Makefile or any other way. Though I believe a situation better than what exists currently is possible.

Reply:
@ScottTodd also `cmake-gui` would not be possible with a `devcontainer`. Since the code can be on a remote machine ... Very classic in most companies. You work on your laptop and connect to a machine over SSH.

Reply:
@ScottTodd I updated the `MakeFile` to include pretty much only "configurations" and point to ready-to-use scripts pre-existing. I hope this illustrates what I was discussing above.

I don't necessarily support the `Makefile` itself, I support a more "easy" way to configure and build.
It could also be writing a `configuration.yaml` and having a script ready this.

I don't care so much about the way we achieve that, more interested in the goal.
And again, the `Makefile` is in a subfolder. Nothing forces you to use it.

Reply:
> Unfortunately nothing as dummy proof / simple as:
> 
> * `make configure`
> * `make build`

That would be either

1. [`build_tools/cmake/build_all.sh`](https://github.com/openxla/iree/blob/main/build_tools/cmake/build_all.sh)
2. [`build_tools/cmake/ctest_all.sh`](https://github.com/openxla/iree/blob/main/build_tools/cmake/ctest_all.sh)

or

1. `cmake -B ../iree-build/ -S .` (or use the defaults for those)
2. `cmake --build ../iree-build/` (note that the default build skips some test deps in the `iree-test-deps` target)
3. `ctest --test-dir ../iree-build/`

We've grown lots of options over time, which keeps making docs like https://openxla.github.io/iree/building-from-source/getting-started/ get more and more complex. Some of those options are just because we depend on LLVM/MLIR for the compiler, some are for user control of which components to build/test, and some are for general toolchain features and quality of life things like caching.

In terms of developer workflows, as project maintainer, I'm not really a fan of having a Makefile that is basically training wheels for CMake. As soon as a developer wants to configure some part of the project they would either need to edit the Makefile (this is a slippery slope into supporting an additional build system) or switch to just using CMake directly. That all being said, if the community gets enough value out of a shared Makefile then we can find a place for it and establish some support level (like how the [Bazel build in LLVM](https://github.com/llvm/llvm-project/tree/main/utils/bazel) is in the ["peripheral support tier"](https://llvm.org/docs/SupportPolicy.html#peripheral-tier)).

> I don't necessarily support the `Makefile` itself, I support a more "easy" way to configure and build.

We were hoping that [CMake "presets"](https://cmake.org/cmake/help/latest/manual/cmake-presets.7.html) would help with this, but they're awkward to use: https://github.com/openxla/iree/issues/5804. I'm also tempted to restructure the [getting started docs](https://openxla.github.io/iree/building-from-source/getting-started/), giving the minimal steps first then having a section on configuration and quality of life tweaks (instead of interleaving those).

---

Concretely for this PR, I think we can make faster progress by reviewing/merging the dev container changes, then we can follow-up with other things like documentation tweaks and the Makefile discussion separately.

Reply:
Sent https://github.com/openxla/iree/pull/13006 to refine the building-from-source/getting-started page. I'd appreciate any feedback you might have :D